Extract method for rendering raw DOT-encoded graphs

fwcd opened the pull request:
This lets users render raw DOT-encoded graphs using the library. Despite maybe being an uncommon use case, it entails few code changes apart from making the renderer slightly more modular.

fwcd committed:
Extract method for rendering raw DOT encoded graphs

This lets users render raw DOT-encoded graphs using the library.

mattt commented on Sources/DOT/Graph+Rendering.swift:
```diff
@@ -29,10 +29,8 @@ fileprivate func which(_ command: String) throws -> URL {
-        let encoded = DOTEncoder().encode(self)
-
+extension String {
+    public func renderAsDOT(using layout: LayoutAlgorithm, to format: Format) throws -> Data {
```
`String` is a commonly used type in Swift, and adding a GraphViz-specific convenience method in an extension has the effect of polluting the API surface area with unrelated functionality.

If the intention is to decouple graph rendering to allow for raw DOT to be rendered by `dot`, then I think a better design would be to create a `Renderer` type that can process both `Graph` and `String` values. The advantage there is that a `Renderer` type could be responsible for formats and layout algorithms, which is currently delegated to `Graph` in a way that arguably conflates model and view concerns.

fwcd replied:
Fixed!

mattt reviewed (changes requested):
Thanks for this PR, @fwcd! I agree that this would be useful to have, and have a slightly different idea for how to approach this.

fwcd committed:
Create a Renderer type

...that handles both graphs and raw DOT-encoded data.

fwcd committed:
Update rendering example in README

fwcd committed:
Update renderer name to 'DOTRenderer' for consistency

mattt commented on Sources/DOT/DOTRenderer.swift:
```diff
@@ -29,10 +29,16 @@ fileprivate func which(_ command: String) throws -> URL {
-    public func render(using layout: LayoutAlgorithm, to format: Format) throws -> Data {
-        let encoded = DOTEncoder().encode(self)
+public struct DOTRenderer {
+    private let layout: LayoutAlgorithm
```
```suggestion
    let layout: LayoutAlgorithm
```

No harm in letting API consumers be able to tell which layout a renderer uses. I could even see a case for making this a `var`.

mattt commented on Sources/DOT/DOTRenderer.swift:
```diff
@@ -29,10 +29,16 @@ fileprivate func which(_ command: String) throws -> URL {
+        self.format = format
+    }
+
+    public func render(dotEncoded: String) throws -> Data {
```
```suggestion
    public init(using layout: LayoutAlgorithm) {
        self.layout = layout
    }

    public func render(dotEncoded: String, to format: Format) throws -> Data {
```

mattt reviewed (commented):
I just realized that I left my review pending this whole time. Submitting this now, in spite of the changes made in the meantime. (They may very well be outdated now).

fwcd committed:
Apply suggestions from code review

Let API consumers access the layout algorithm and
move format parameter to render method.

Co-authored-by: Mattt <mattt@me.com>

fwcd committed:
Make the layout algorithm public

fwcd committed:
Add missing format parameter to DOTRenderer.render(graph:)

mattt committed:
Add Tools module

Move Renderer and which helper method to Tools module

Restore Graph rendering extension

mattt committed:
Allow renderer to be initialized directly with tool URL

mattt committed:
Remove conditional for Swift >= 5.2

mattt committed:
Enable rendering tests on Linux

mattt committed:
Add missing DOT module dependency for Tools module

mattt committed:
Install system dependencies on macOS

mattt committed:
Refactor error handling in rendering tests

mattt committed:
Add documentation comments for rendering APIs

mattt committed:
Add changelog entry for #5

mattt commented:
Thanks again for your work on this and your patience. I went ahead and made the changes I wanted for this feature. These will be available in the forthcoming 0.2.0 release. Let me know if you have any questions or feedback on these. Thanks!
